src/Interpreters/ColumnAliasesVisitor.cpp

```diff
@@ -81,13 +61,25 @@ void ColumnAliasesMatcher::visit(ASTIdentifier & node, ASTPtr & ast, Data & data
 {
     if (auto column_name = IdentifierSemantic::getColumnName(node))
     {
-        if (data.forbidden_columns.count(*column_name) || data.private_aliases.count(*column_name) || !data.columns.has(*column_name))
+        if (data.array_join_result_columns.count(*column_name) || data.array_join_source_columns.count(*column_name)
+            || data.private_aliases.count(*column_name) || !data.columns.has(*column_name))
             return;
 
         const auto & col = data.columns.get(*column_name);
         if (col.default_desc.kind == ColumnDefaultKind::Alias)
         {
+            auto alias = node.tryGetAlias();
+            auto alias_expr = col.default_desc.expression->clone();
+            auto original_column = alias_expr->getColumnName();
+            // If expanded alias is used in array join, avoid expansion, otherwise the column will be mis-array joined
+            if (data.array_join_result_columns.count(original_column) || data.array_join_source_columns.count(original_column))
+                return;
             ast = addTypeConversionToAST(col.default_desc.expression->clone(), col.type->getName(), data.columns.getAll(), data.context);
```

Comment:
Looks like we can use `alias_expr` here.

Reply:
Definitely)